runner/src/Quickstrom/Run.hs

```diff
@@ -398,26 +423,42 @@ focus =
     Just e -> tryAction (ActionSuccess <$ (elementSendKeys "" e))
     Nothing -> pure ActionImpossible
 
-runAction :: WebDriver m => Action Selected -> Runner m ActionResult
-runAction = \case
+runBaseAction :: (MonadIO m, WebDriver m) => BaseAction Selected -> Runner m ActionResult
+runBaseAction = \case
   Focus s -> focus s
   KeyPress c -> sendKey c
   EnterText t -> sendKeys t
   Click s -> click s
+  Await s -> awaitElement defaultAwaitSecs s
+  AwaitWithTimeoutSecs i s -> awaitElement i s
   Navigate uri -> tryAction (ActionSuccess <$ navigateTo uri)
 
+runAction :: (MonadIO m, WebDriver m) => SelectedActionSequence -> Runner m ActionResult
+runAction = \case
+    [] -> do pure ActionImpossible
+    (h:[]) -> runBaseAction h
+    (h:t) -> recurse t $ runBaseAction h
+  where
+    recurse t r =
+      do res <- r
+         case res of
+           ActionFailed _ -> r
+           ActionImpossible -> r
```

Comment:
It looks like these would rerun the failed action, rather than just returning `res`?

Reply:
So, it should be a `pure res` instead? I was a little unsure and just let the types guide me here.  `Runner m ActionResult` (in a monadic context) is required so I figured the value _before_ it was extracted from its monadic context would be what’s required by the compiler.

Reply:
i.e., to be honest I don’t completely understand the `Runner` monad and relied on an admittedly _mechanistic_ interpretation in some places (rather than a true semantic understanding).

Reply:
Yeah, just `pure res`. :+1: